Ratelimit validation (#1508)
kyma-project/api-gateway: docs/user/custom-resources/ratelimit/04-00-ratelimit.md

RateLimit Controller uses to configure the request rate limit for applications.

To get the up-to-date CRD in the `yaml` format, run the following command:

```shell
kubectl get crd ratelimits.gateway.kyma-project.io -o yaml
```

## Specification

| Field                                 | Required | Description                                                                                                                                                                                                                                                                                                                                                                                        |
|---------------------------------------|----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **selectorLabels**                    | **YES**  | Labels that specify the set of Pods to which the configuration applies.<br/> Each Pod can match only one RateLimit CR.<br/> The label scope is limited to the namespace where the resource is located.                                                                                                                                                                                             |

Reviewer: ```suggestion
| **selectorLabels**                    | **YES**  | Labels that specify the set of Pods to which the configuration applies.<br/> Each Pod must match only one RateLimit CR.<br/> The label scope is limited to the namespace where the resource is located.                                                                                                                                                                                             |
```

| **local**                             | **YES**  | Local rate limit configuration.                                                                                                                                                                                                                                                                                                                                                                    |
| **local.defaultBucket**               | **YES**  | The default token bucket for rate limiting requests. <br/>If additional local buckets are configured in the same RateLimit CR, this bucket serves as a fallback for requests that don't match any other bucket's criteria. <br/>Each request consumes a single token. If a token is available, the request is allowed. If no tokens are available, the request is rejected with status code `429`. |

> | **local.defaultBucket.maxTokens**     | **YES**  | The maximum tokens that the bucket can hold. This is also the number of tokens that the bucket initially contains.                                                                                                                                                                                                                                                                                 |

Reviewer: ```suggestion
| **local.defaultBucket.maxTokens**     | **YES**  | The maximum number of tokens that the bucket can hold. This is also the number of tokens that the bucket initially contains.                                                                                                                                                                                                                                                                                 |
```

| **local.defaultBucket.tokensPerFill** | **YES**  | The number of tokens added to the bucket during each fill interval.                                                                                                                                                                                                                                                                                                                                |

> | **local.defaultBucket.fillInterval**  | **YES**  | The fill interval that tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket will never contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   |

Reviewer: ```suggestion
| **local.defaultBucket.fillInterval**  | **YES**  | The fill interval during which tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket will never contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   |
```

> | **local.buckets**                     | **NO**   | List of additional rate limit buckets for requests. <br/>Each bucket must specify either a `path` or `headers`. <br/>For each request matching the bucket's criteria, a single token is consumed. If a token is available, the request is allowed. If no tokens are available, the request is rejected with status code `429`.                                                                     |

Reviewer: ```suggestion
| **local.buckets**                     | **NO**   | Specifies a list of additional rate limit buckets for requests. <br/>Each bucket must specify either a `path` or `headers`. <br/>For each request matching the bucket's criteria, a single token is consumed. If a token is available, the request is allowed. If no tokens are available, the request is rejected with status code `429`.                                                                     |
```

| **local.buckets.path**                | **NO**   | Specifies the path to be rate limited starting with `/`. <br/>For example, `/foo`.                                                                                                                                                                                                                                                                                                                 |
| **local.buckets.headers**             | **NO**   | Specifies the request headers to be rate limited. The key is the header name, and the value is the header value. All specified headers must be present in the request for this configuration to match. For example, `x-api-usage: BASIC`.                                                                                                                                                          |                                                                                                                                                                                    

> | **local.buckets.maxTokens**           | **YES**  | The maximum tokens that the bucket can hold. This is also the number of tokens that the bucket initially contains.                                                                                                                                                                                                                                                                                 |

Reviewer: ```suggestion
| **local.buckets.maxTokens**           | **YES**  | The maximum number of tokens that the bucket can hold. This is also the number of tokens that the bucket initially contains.                                                                                                                                                                                                                                                                                 |
```

| **local.buckets.tokensPerFill**       | **YES**  | The number of tokens added to the bucket during each fill interval.                                                                                                                                                                                                                                                                                                                                |

> | **local.buckets.fillInterval**        | **YES**  | The fill interval that tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket will never contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   |

Reviewer: ```suggestion
| **local.buckets.fillInterval**        | **YES**  | The fill interval that tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket cannot contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   |
```
